pkg/vulnloader/nvdloader/loader.go

```diff
@@ -1,139 +1,249 @@
 	vulnloader.RegisterLoader(vulndump.NVDDirName, &loader{})
 }
 
+var _ vulnloader.Loader = (*loader)(nil)
+
 type loader struct{}
 
 // DownloadFeedsToPath downloads the NVD feeds to the given path.
 // If this function is successful, it will fill the directory with
 // one json file for each year of NVD data.
 func (l *loader) DownloadFeedsToPath(outputDir string) error {
 	// Fetch NVD enrichment data from curated repos
-	enrichmentMap, err := Fetch()
+	enrichments, err := Fetch()
 	if err != nil {
-		return errors.Wrap(err, "could not fetch NVD enrichment sources")
+		return fmt.Errorf("could not fetch NVD enrichment sources: %w", err)
 	}
 
 	nvdDir := filepath.Join(outputDir, vulndump.NVDDirName)
 	if err := os.MkdirAll(nvdDir, 0755); err != nil {
-		return errors.Wrapf(err, "creating subdir for %s", vulndump.NVDDirName)
+		return fmt.Errorf("creating subdir for %s: %w", vulndump.NVDDirName, err)
 	}
-	endYear := time.Now().Year()
-	for year := 2002; year <= endYear; year++ {
-		if err := downloadFeedForYear(enrichmentMap, nvdDir, year); err != nil {
-			return err
-		}
+
+	var fileNo, totalVulns int
+
+	// Explicitly set startIdx to parallel how this is all done within the loop below.
+	startIdx := 0
+	apiResp, err := query(fmt.Sprintf(urlFmt, startIdx))
+	if err != nil {
+		return err
 	}
-	return nil
-}
+	var i int
+	// Buffer to store vulns until they are written to a file.
+	cveItems := make([]*jsonschema.NVDCVEFeedJSON10DefCVEItem, 0, 20_000)
+	for apiResp.ResultsPerPage != 0 {
+		vulns, err := toJSON(apiResp.Vulnerabilities)
+		if err != nil {
+			return fmt.Errorf("failed to convert API vulns to JSON: %w", err)
+		}
 
-func removeInvalidCPEs(item *schema.NVDCVEFeedJSON10DefNode) {
-	cpeMatches := item.CPEMatch[:0]
-	for _, cpeMatch := range item.CPEMatch {
-		if cpeMatch.Cpe23Uri == "" {
-			cpeMatches = append(cpeMatches, cpeMatch)
-			continue
+		if len(vulns) != 0 {
+			cveItems = append(cveItems, vulns...)
+
+			i++
+			// Write to disk every ~20,000 vulnerabilities.
+			if i == 10 {
+				i = 0
+
+				enrichCVEItems(&cveItems, enrichments)
+
+				feed := &jsonschema.NVDCVEFeedJSON10{
+					CVEItems: cveItems,
+				}
+				if err := writeFile(filepath.Join(nvdDir, fmt.Sprintf("%d.json", fileNo)), feed); err != nil {
+					return fmt.Errorf("writing to file: %w", err)
+				}
+
+				fileNo++
+				totalVulns += len(cveItems)
+				log.Infof("Loaded %d NVD vulnerabilities", totalVulns)
+				// Reduce, reuse, and recycle.
+				cveItems = cveItems[:0]
+			}
 		}
-		attr, err := wfn.UnbindFmtString(cpeMatch.Cpe23Uri)
+
+		// Rudimentary rate-limiting.
+		// NVD limits users without an API key to roughly one call every 6 seconds.
+		// With an API key, it is roughly one call every 0.6 seconds.
+		// At this time, the API key is inaccessible in CI, so we opt to play it safe and make a call every 10 seconds.
+		// As of writing there are ~216,000 vulnerabilities, so this whole process should take ~18 minutes.
+		time.Sleep(10 * time.Second)
```

Comment:
Just curious, how long does it approximately take to get all data?

Reply:
about 18 minutes. Says it on the line directly above this 😄

Reply:
It's odd that it was taking longer in scanner v4 workflow with not API key